wrapper/src/main/java/software/amazon/jdbc/WeightedRandomHostSelector.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc;
+
+import java.sql.SQLException;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Properties;
+import java.util.Random;
+import java.util.concurrent.Callable;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import software.amazon.jdbc.hostavailability.HostAvailability;
+import software.amazon.jdbc.util.Messages;
+
+public class WeightedRandomHostSelector implements HostSelector {
+  public static final AwsWrapperProperty WEIGHTED_RANDOM_HOST_WEIGHT_PAIRS = new AwsWrapperProperty(
+      "weightedRandomHostWeightPairs", null,
+      "Comma separated list of database host-weight pairs in the format of `<host>:<weight>`.");
+  public static final String STRATEGY_WEIGHTED_RANDOM = "weightedRandom";
+  static final int DEFAULT_WEIGHT = 1;
+  static final Pattern HOST_WEIGHT_PAIRS_PATTERN =
+      Pattern.compile("((?<host>[^:/?#]*):(?<weight>[0-9]*))");
+
+  private Map<String, Integer> cachedHostWeightMap;
+  private String cachedHostWeightMapString;
+  private Random random;
+  private Callable<Integer> randomFunc;
+
+  public HostSpec getHost(
+      @NonNull List<HostSpec> hosts,
+      @NonNull HostRole role,
+      @Nullable Properties props) throws SQLException {
+
+    final Map<String, Integer> hostWeightMap =
+        this.getHostWeightPairMap(WEIGHTED_RANDOM_HOST_WEIGHT_PAIRS.getString(props));
+
+    // Get and check eligible hosts
+    final List<HostSpec> eligibleHosts = hosts.stream()
+        .filter(hostSpec ->
+            role.equals(hostSpec.getRole()) && hostSpec.getAvailability().equals(HostAvailability.AVAILABLE))
+        .sorted(Comparator.comparing(HostSpec::getHost))
+        .collect(Collectors.toList());
+
+    if (eligibleHosts.isEmpty()) {
+      throw new SQLException(Messages.get("HostSelector.noHostsMatchingRole", new Object[] {role}));
+    }
+
+    final Map<HostSpec, NumberRange> hostWeightRangeMap = new HashMap<>();
```

Comment:
Something just occurred to me, I've read that mutable objects should not be used as keys in hash sets or hash maps because their hashcode changes when the object changes which can lead to lost entries, inconsistent lookup results etc. I looked in our code and I think there are some places where we use mutable keys, as a separate task we should have someone look through the code and clean these up

Reply:
Oh thats a good one. Thanks for pointing that out. Will fix.